### PR TITLE
[14663] Use Mersenne Twister as random generator

### LINKED ISSUE #2694: Program infinitely hangs while generating random value on AMD Ryzen 3600.
### Is there an already existing issue for this?

- [X] I have searched the existing issues

### Expected behavior

The `RTPSParticipant` object is created immediately on AMD Ryzen 3600 hardware.

### Current behavior

While creating `RTPSParticipant` with `RTPSDomain::createParticipant(...)`, the program infinitely hangs on AMD Ryzen 5 3600 hardware. Line 107 of the code below is where the code hangs.
[https://github.com/eProsima/Fast-DDS/blob/6a978a70589074f700d9b7c00cbaf91ba43d5031/src/cpp/rtps/common/GuidUtils.hpp#L105-L107](https://github.com/eProsima/Fast-DDS/blob/6a978a70589074f700d9b7c00cbaf91ba43d5031/src/cpp/rtps/common/GuidUtils.hpp#L105-L107)

While investigating the issue, I found the following posts.
- [https://arstechnica.com/gadgets/2019/10/how-a-months-old-amd-microcode-bug-destroyed-my-weekend/](https://arstechnica.com/gadgets/2019/10/how-a-months-old-amd-microcode-bug-destroyed-my-weekend/)
- https://gitlab.com/gromacs/gromacs/-/issues/3855
- https://github.com/systemd/systemd/issues/11810
- https://github.com/systemd/systemd/issues/18184

According to the posts, the issue seems to be a bug due to a random number generator of AMD 3000 series CPUs.

### Steps to reproduce

The identical code block from the `RTPSDomain.cpp` reproduces  the identical issue on Ryzen 3600 cpu.
The program hangs infinitely on the line `uint16_trand_value = distribution(generator);`
```c++
#include <random>

int main() {
  std::random_device generator;
  std::uniform_int_distribution<uint16_t> distribution(0, std::numeric_limits<uint16_t>::max());
  uint16_t rand_value = distribution(generator);
}
```

### Fast DDS version/commit

Tested the codes in the version which ros2 foxy version depends on. The code remains in the latest branch, too.

tag: 2.1.x
commit: 4db32122183390f895138bd569c6fa847d80b2c7

### Platform/Architecture

Ubuntu Focal 20.04 amd64

### Transport layer

Default configuration, UDPv4 & SHM

### Additional context

I guess that adding `std::mt19937` generator could help to prevent the infinite hanging issue even in case that the `std::random_device` does not work properly (it may not guarantee the random performance of the case).

```cpp
#include <random>

int main() {
  std::random_device rd;
  std::mt19937 generator(rd());
  std::uniform_int_distribution<uint16_t> distribution(0, std::numeric_limits<uint16_t>::max());
  uint16_t rand_value = distribution(generator);
}
```

### XML configuration file

_No response_

### Relevant log output

_No response_

### Network traffic capture

_No response_

### PR DESCRIPTION
**Pull request by MiguelCompany**
Signed-off-by: Miguel Company <MiguelCompany@eprosima.com>

## Description

This should fix #2694 by using a Mersenne Twister RNG seeded by the random device. Since the GUID of the participant also includes the process identifier, the uniqueness of the GUID will still be enough on most use-cases.

@mergifyio backport 2.5.x 2.3.x 2.1.x

## Contributor Checklist
- [X] Commit messages follow the project guidelines. 
- [X] The code follows the style guidelines of this project. 
- **N/A** Tests that thoroughly check the new feature have been added/Regression tests checking the bug and its fix have been added. 
- **N/A** Any new/modified methods have been properly documented using Doxygen. 
- [ ] Fast DDS test suite has been run locally. 
- [X] Changes are ABI compatible. 
- [X] Changes are API compatible. 
- **N/A** Documentation builds and tests pass locally. 
- **N/A** New feature has been added to the `versions.md` file (if applicable).
- **N/A** New feature has been documented/Current behavior is correctly described in the documentation. 

## Reviewer Checklist
- [ ] Check contributor checklist is correct.
- [ ] Check CI results: changes do not issue any warning.
- [ ] Check CI results: failing tests are unrelated with the changes.

### CODE REVIEW

**Comment by MiguelCompany** [comment]
@VladimirVassiliev Since you commented on #1637, where the random generator was added, I kindly ask for your opinion on this one.

File: src/cpp/rtps/common/GuidUtils.hpp [hunk 1]
```diff
@@ -102,7 +102,8 @@ class GuidUtils
         prefix_.value[4] = static_cast<octet>(pid & 0xFF);
         prefix_.value[5] = static_cast<octet>((pid >> 8) & 0xFF);
 
-        std::random_device generator;
+        std::random_device device;
```
Comment (richiware) [inline src/cpp/rtps/common/GuidUtils.hpp:105]: Maybe a comment may help in the future, the reason of using this random generator or maybe the number of the issue.

**Review by richiware — CHANGES_REQUESTED** [review]

File: src/cpp/rtps/common/GuidUtils.hpp (hunk 1 above)
Comment (MiguelCompany) [reply src/cpp/rtps/common/GuidUtils.hpp:108]: Done

**Review by richiware — APPROVED** [review]

**Comment by VladimirVassiliev** [comment]
How does this fix the problem if random_device is still being called? The real time clock could be used to seed, if available. It may be the best to allow the user to supply a unique enough value (like boot counter or platform specific non-deterministic random generator) in cases when random_device does not work.

**Comment by EduPonz** [comment]
We have decided to close this PR without merging it as the bug has been solved at the origin. In any case, this will stay here for future reference. Thanks everyone!

**Pull request closed without merging**